src/libraries/Common/src/System/Collections/Generic/EnumerableHelpers.cs

```diff
@@ -8,6 +8,30 @@ namespace System.Collections.Generic
     /// </summary>
     internal static partial class EnumerableHelpers
     {
+        /// <summary>Attempt to determine the count of the source enumerable without forcing an enumeration.</summary>
+        /// <param name="source">The source enumerable.</param>
+        /// <param name="count">The count determined by the type test.</param>
+        /// <returns>
+        /// True if the source enumerable could be determined without enumerating, false otherwise.
+        /// </returns>
+        internal static bool TryGetCount<T>(IEnumerable<T> source, out int count)
+        {
+            if (source is ICollection<T> ict)
+            {
+                count = ict.Count;
+                return true;
+            }
+
+            if (source is ICollection ic)
+            {
+                count = ic.Count;
+                return true;
+            }
```

Comment:
FWIW, it's a little weird to me that we now have an inconsistency where the ctor uses EnumerableHelpers.ToArray which doesn't check ICollection but here we are checking ICollection.

Reply:
The `ToArray` method additionally depends on `ICollection<T>.CopyTo`. I could duplicate the logic for `ICollection` however using the untyped `ICollection.CopyTo(Array)` method felt wrong to me in this context. I think we could simply use `ICollection.Count` to presize the target buffer:
```csharp
const int DefaultCapacity = 4;
T[] arr = new T[source is ICollection icu ? icu.Count : DefaultCapacity];
```

Reply:
In the end I removed `ICollection` detection from the class altogether.